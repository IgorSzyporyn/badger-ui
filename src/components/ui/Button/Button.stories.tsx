import React from 'react'
import { withTests } from '@storybook/addon-jest'
import { withDesign } from 'storybook-addon-designs'
import { withPerformance } from 'storybook-addon-performance'
import { Button } from './Button'
import results from '../../../../.jest-test-results.json'

import type { Story, Meta } from '@storybook/react'
import type { ButtonProps } from './Button'

export default {
  title: 'UI Components/Button',
  component: Button,
  decorators: [withDesign, withTests({ results }), withPerformance],
  parameters: {
    actions: { handles: ['click', 'dblclick'] },
    layout: 'centered',
    options: { showPanel: true },
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/yl44JcFlLYJ1IHwE6JzbUl/Level-2.-Rollout?node-id=601%3A1018',
    },
  },
} as Meta

const Template: Story<ButtonProps> = (args) => <Button {...args} />

export const Controllable = Template.bind({})

Controllable.args = {
  label: 'Button',
}

Controllable.parameters = {
  jest: ['Button.test.tsx'],
}
