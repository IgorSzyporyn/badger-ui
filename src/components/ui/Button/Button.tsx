import React, { useRef } from 'react'
import { nanoid } from 'nanoid'
import {
  FilledButton,
  OutlinedButton,
  MutedButton,
  FlatButton,
  ButtonInner,
  Adornment,
  Label,
} from './styled'
import { Icon } from '../Icon/Icon'

import type {
  ThemeAdorment,
  ThemeAdormentType,
  ThemeColorType,
  ThemeElevationType,
  ThemeSizeType,
  ThemeVariant,
} from '../../../theme/types'
import type { IconProps, IconType } from '../Icon/Icon'

const ButtonVariantMap = {
  filled: FilledButton,
  outlined: OutlinedButton,
  muted: MutedButton,
  flat: FlatButton,
}

export type ButtonProps = {
  elevation?: ThemeElevationType
  adornment?: ThemeAdorment
  busy?: boolean
  disabled?: boolean
  fullWidth?: boolean
  icon?: IconType
  iconPosition?: ThemeAdormentType
  label?: string
  size?: ThemeSizeType
  type?: ThemeColorType
  variant?: ThemeVariant
  AdornmentProps?: React.HTMLAttributes<HTMLDivElement>
  IconProps?: IconProps
  LabelProps?: React.HTMLAttributes<HTMLDivElement>
} & React.HTMLAttributes<HTMLDivElement>

export const Button = React.forwardRef(
  (
    {
      adornment = {},
      elevation = 0,
      busy = false,
      children,
      disabled = false,
      fullWidth = false,
      icon,
      iconPosition = 'start',
      label,
      size = 'normal',
      type = 'primary',
      variant = 'filled',
      AdornmentProps = {},
      IconProps = {},
      LabelProps = {},
      ...rest
    }: ButtonProps,
    ref: React.Ref<HTMLDivElement>
  ) => {
    const idRef = useRef(nanoid())
    const id = idRef.current

    const ButtonVariant = ButtonVariantMap[variant]

    return (
      <ButtonVariant
        data-cylindoui-button=""
        data-cylindoui-variant={variant}
        data-cylindoui-type={type}
        id={id}
        type={type}
        disabled={disabled}
        variant={variant}
        _elevation={elevation}
        fullWidth={fullWidth}
        size={size}
        role="button"
        {...rest}
        ref={ref}
      >
        <ButtonInner data-cylindoui-button-inner="" size={size}>
          {icon && iconPosition === 'start' && (
            <Adornment
              data-cylindoui-button-adornment=""
              data-cylindoui-button-adornment-start=""
              {...AdornmentProps}
            >
              <Icon
                data-cylindoui-button-icon=""
                icon={icon}
                size={size}
                {...IconProps}
              />
            </Adornment>
          )}
          {adornment.start !== undefined && (
            <Adornment
              data-cylindoui-button-adornment=""
              data-cylindoui-button-adornment-start=""
              {...AdornmentProps}
            >
              {adornment.start}
            </Adornment>
          )}
          <Label data-cylindoui-button-label="" {...LabelProps}>
            {label ? label : children}
          </Label>
          {adornment.end !== undefined && (
            <Adornment
              data-cylindoui-button-adornment=""
              data-cylindoui-button-adornment-end=""
              {...AdornmentProps}
            >
              {adornment.end}
            </Adornment>
          )}
          {icon && iconPosition === 'end' && (
            <Adornment
              data-cylindoui-button-adornment=""
              data-cylindoui-button-adornment-end=""
              {...AdornmentProps}
            >
              <Icon
                data-cylindoui-button-icon=""
                icon={icon}
                size={size}
                {...IconProps}
              />
            </Adornment>
          )}
        </ButtonInner>
      </ButtonVariant>
    )
  }
)
