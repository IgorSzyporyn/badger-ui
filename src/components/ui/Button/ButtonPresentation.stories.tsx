import React from 'react'
import { ButtonPreview } from '../../storybook/ButtonPreview/ButtonPreview'

import type { Story, Meta } from '@storybook/react'

export default {
  title: 'UI Components/Button/Presentation',
  parameters: {
    layout: 'centered',
    controls: { disable: true },
    actions: { disable: true },
    a11y: { disable: true },
    test: { disable: true },
    performance: { disable: true },
    options: {
      showPanel: false,
    },
    previewTabs: {
      'storybook/docs/panel': { hidden: true },
    },
  },
} as Meta

export const Presentation: Story = () => <ButtonPreview />
