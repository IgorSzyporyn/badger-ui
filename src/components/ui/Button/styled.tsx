import styled, { css } from 'styled-components'

import type {
  ThemeColorType,
  ThemeElevationType,
  ThemeSizeType,
  ThemeVariant,
} from '../../../theme/types'

const heightAdjuster = 2.8
const paddingAdjuster = 2

type WrapperProps = {
  disabled: boolean
  fullWidth: boolean
  size: ThemeSizeType
  type: ThemeColorType
  _elevation: ThemeElevationType
  variant: ThemeVariant
}

const BaseButton = styled.div<WrapperProps>`
  background-color: transparent;
  border: 1px solid transparent;
  cursor: pointer;
  font-weight: 500;
  text-transform: uppercase;
  transition: 'box-shadow 125ms ease-in-out';

  ${({ fullWidth, theme, _elevation, size, disabled }) => css`
    display: ${fullWidth ? 'block' : 'inline-block'};
    border-radius: ${theme && theme.metrics ? theme.metrics.borderRadius : ''};
    font-size: ${theme && theme.size ? `${theme.size.font[size]}px` : ''};
    opacity: ${disabled ? 0.4 : 1};
    box-shadow: ${_elevation && theme && theme.elevations
      ? theme.elevations[_elevation]
      : ''};
  `}
`

export const FilledButton = styled(BaseButton)<WrapperProps>`
  ${({ theme, type }) => css`
    background-color: ${theme && theme.color ? theme.color[type].normal : ''};
    color: ${theme && theme.color ? theme.color[type].text : ''};
  `}
`

export const OutlinedButton = styled(BaseButton)<WrapperProps>`
  ${({ theme, type }) => css`
    border-color: ${theme && theme.color ? theme.color[type].normal : ''};
    color: ${theme && theme.color ? theme.color[type].normal : ''};
  `}
`

export const MutedButton = styled(BaseButton)<WrapperProps>`
  ${({ theme, type }) => css`
    background-color: ${theme && theme.color ? theme.color[type].darker : ''};
    border-color: ${theme && theme.color ? theme.color[type].darker : ''};
    color: ${theme && theme.color ? theme.color[type].lightest : ''};
  `}
`

export const FlatButton = styled(BaseButton)<WrapperProps>`
  color: ${({ theme, type }) => (theme && theme.color ? theme.color[type].normal : '')};
`

type ButtonInnerProps = {
  size: ThemeSizeType
}

export const ButtonInner = styled.div<ButtonInnerProps>`
  display: flex;
  align-items: center;
  justify-content: center;

  ${({ theme, size }) => {
    const fontSize = theme && theme.size && theme.size.font[size]

    return css`
      height: ${Math.round(fontSize * heightAdjuster)}px;
      padding: 0 ${Math.round(fontSize * paddingAdjuster)}px;
    `
  }}
`

export const Adornment = styled.div`
  display: flex;

  &[data-cylindoui-button-adornment-start] {
    margin-right: 0.5em;

    &:last-child {
      margin-right: 0;
    }
  }

  &[data-cylindoui-button-adornment-end] {
    margin-left: 0.5em;

    &:last-child {
      margin-left: 0;
    }
  }
`

export const Label = styled.div`
  line-height: 1;
`
