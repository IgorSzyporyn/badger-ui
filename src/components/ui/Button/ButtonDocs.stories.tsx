import React from 'react'
import { Button } from './Button'

import type { Story, Meta } from '@storybook/react'
import type { ButtonProps } from './Button'

export default {
  title: 'UI Components/Button/Documentation',
  component: Button,
  parameters: {
    docsOnly: true,
    facelift: {
      docs: {
        type: 'simple',
      },
    },
  },
} as Meta

const Template: Story<ButtonProps> = (args) => <Button {...args} />

export const Documentation = Template.bind({})
Documentation.args = {}
