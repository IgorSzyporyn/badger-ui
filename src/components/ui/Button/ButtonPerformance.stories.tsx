import React from 'react'
import { Story, Meta } from '@storybook/react'
import { Button } from './Button'

import type { ButtonProps } from './Button'

export default {
  title: 'UI Components/Button/Performance',
  component: Button,
  parameters: {
    layout: 'centered',
    controls: { disabled: true },
    actions: { disabled: true },
    a11y: { disabled: true },
    options: { showPanel: false },
    previewTabs: {
      'storybook/docs/panel': { hidden: true },
    },
  },
} as Meta

const array1000 = [...Array(5000).keys()]
const array5000 = [...Array(5000).keys()]
const array10000 = [...Array(5000).keys()]

export const Render1000Buttons: Story<ButtonProps> = () => (
  <div
    style={{
      display: 'grid',
      gridTemplateColumns: '1fr 1fr 1fr 1fr 1fr 1fr 1fr',
      gridGap: 8,
    }}
  >
    {array1000.map((i) => (
      <Button key={`Button-Render-Test-${i}`} label={`#${i}`} />
    ))}
  </div>
)

export const Render5000Buttons: Story<ButtonProps> = () => (
  <div
    style={{
      display: 'grid',
      gridTemplateColumns: '1fr 1fr 1fr 1fr 1fr 1fr 1fr',
      gridGap: 8,
    }}
  >
    {array5000.map((i) => (
      <Button key={`Button-Render-Test-${i}`} label={`#${i}`} />
    ))}
  </div>
)

export const Render10000Buttons: Story<ButtonProps> = (args) => (
  <div
    style={{
      display: 'grid',
      gridTemplateColumns: '1fr 1fr 1fr 1fr 1fr 1fr 1fr',
      gridGap: 8,
    }}
  >
    {array10000.map((i) => (
      <Button key={`Button-Render-Test-${i}`} label={`#${i}`} {...args} />
    ))}
  </div>
)
