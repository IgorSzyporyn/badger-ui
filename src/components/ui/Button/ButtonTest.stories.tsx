import React from 'react'
import { Meta } from 'storybook-facelift'
import results from '../../../../.jest-test-results.json'
import { withTests } from '@storybook/addon-jest'
import { Button } from './Button'

export default {
  title: 'UI Components/Button/Test',
  component: Button,
  parameters: {
    layout: 'centered',
    controls: { disabled: true },
    actions: { disabled: true },
    a11y: { disabled: true },
    decorators: [withTests({ results })],
    options: {
      showPanel: true,
    },
    previewTabs: {
      'storybook/docs/panel': { hidden: true },
    },
    facelift: {
      addProvider: false,
    },
  },
} as Meta

export const Test = () => <div>Jest results in storybook</div>

Test.parameters = {
  jest: ['./Button.test.tsx'],
}
