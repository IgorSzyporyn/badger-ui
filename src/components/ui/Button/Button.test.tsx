import React from 'react'
import { cleanup, render } from '@testing-library/react'
import { Button } from './Button'

describe('<Button />', () => {
  it('can perform a shallow render', () => {
    const { findByTestId } = render(<Button data-test-id="button" />)

    expect(findByTestId('button')).toBeDefined()

    cleanup()
  })

  it('will render correctly with minimum properties', () => {
    const { findByTestId, findByRole } = render(
      <Button data-test-id="button" label="Button Label" />
    )

    expect(findByTestId('button')).toBeDefined()
    expect(findByRole('button')).toBeDefined()

    cleanup()
  })

  it('can match snapshot for a shallow render', () => {
    const { container } = render(<Button />)
    expect(container).toMatchSnapshot()

    cleanup()
  })
})
