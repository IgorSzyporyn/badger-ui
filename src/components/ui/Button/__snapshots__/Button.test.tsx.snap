// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> can match snapshot for a shallow render 1`] = `
<div>
  <div
    class="styled__BaseButton-bhq7k-0 styled__FilledButton-bhq7k-1 IbMUB hcxkLz"
    data-cylindoui-button=""
    data-cylindoui-type="primary"
    data-cylindoui-variant="filled"
    id="bw4Y-i5vzZHT0_-G47Wxk"
    role="button"
    type="primary"
  >
    <div
      class="styled__ButtonInner-bhq7k-5 cDIKif"
      data-cylindoui-button-inner=""
    >
      <div
        class="styled__Label-bhq7k-7 jiWxsY"
        data-cylindoui-button-label=""
      />
    </div>
  </div>
</div>
`;
